pandas/core/internals/concat.py

```diff
@@ -443,7 +452,7 @@ def _is_uniform_join_units(join_units: List[JoinUnit]) -> bool:
     #  cannot necessarily join
     return (
         # all blocks need to have the same type
-        all(isinstance(ju.block, type(join_units[0].block)) for ju in join_units)
+        all(type(ju.block) is type(join_units[0].block) for ju in join_units)
```

Comment:
this change is to ensure DatetimeBlock and DatetimeTZBlock don't incorrectly take this "uniform join units" code path